### PR TITLE
Add Equatable conformance to Graph and Edge

### PR DESCRIPTION
**Pull request by ferranpujolcamins**
As discussed [in this issue](https://github.com/davecom/SwiftGraph/issues/35), this is a first step towards a SetAlgebra conformant subclass of Graph. SetAlgebra inherits from Equatable, so Equatable needed to be implemented. Since it makes sense to check for graph equality, I implemented it for Graph basse class.

When a protocol inherits from Equatable it cannot be used as type anymore. This was a problem with Edge. To solve it, a type erased AnyEquatableEdge is created. See [this link](https://khawerkhaliq.com/blog/swift-protocols-equatable-part-two/) for details.

### CODE REVIEW

**Comment by ferranpujolcamins** [comment]
Well, I think we might refuse to conform to SetAlgebra because it inherits from [ExpressibleByArrayLiteral](https://developer.apple.com/documentation/swift/expressiblebyarrayliteral). I don't see any need for ExpressibleByArrayLiteral, on the contrary it'd a pain to implement. Why is it even needed for SetAlgebra?
Furthermore, the only advantage of conformance to SetAlgebra is the default implementations it provides. We can implement the methods we want ourselves, it's not that big deal.

Anyway, this PR is relevant by itself I think. I just wanted to write my thoughts about SetAlgebra.

**Comment by davecom** [comment]
Hi @ferranpujolcamins,
Thanks for the work on this. I don't think it makes sense to merge it though. You made a type-erased struct that implements `Edge` to enable `Equatable`, but both of the concrete class implementations of `Edge`, `UnweightedEdge` and `WeightedEdge`, are already `Equatable`. I also am not sure if it makes sense for the protocol itself to implement `Equatable` since it will inherently differ comparing `UnweightedEdge`s and `WeightedEdge`s, so I think it makes more sense for `Equatable` support to live "a level down" as it already does. I agree with you, I don't think that implementing `SetAlgebra` makes a ton of sense as-is, but I need to explore the documentation you linked to more carefully.

Basically I feel like we're adding more complexity to the project than we're necessarily getting benefit. If the goal is to create a subclass of `Graph` that we can do set operations on for things like Graph Union (#35), then why not simply override adding vertices to ensure there are never any duplicates (as I think you've done before) in the subclass and then we can implement all of the set operations on that `Graph` subclass assuming we have an array of only unique elements. Does that make sense? I think it would be a lot simpler and would restrict the changes to a single a subclass.

Thanks,
David

**Comment by ferranpujolcamins** [comment]
Hi Dave, yes, this PR has nothing to do with set operations. I started becasue at the beggining I was going for SetAlgebra conformance, and SetAlgebra inherits from Equatable. Furthermore, it makes sense to me for Graph to implement Equatable, although it is true there's no clear use case for it.

I understand that there's no clear use case for value semantics graph equality that justifies the complexity of this PR. However, if we ever do this, there's no alternative to make Edge equatable in a way or another.

Thank you for reviewing this!
